### PR TITLE
Support AT TIME ZONE

### PR DESCRIPTION
**Pull request by hash-16**
### Description

It converts an inputdate (datetime,smalldatetime,datetime2,datetimeoffset) to the corresponding datetimeoffset value in the target time zone.

```
Select convert(datetime,'2022-03-27T01:01:00') AT TIME ZONE 'Atlantic Standard Time'

2022-03-27 01:01:00.000 -03:00 

Here it considers that the given time is currently in the target time zone so no change in the time is observed
it just adds the offset.
```
Engine PR :- [Link](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/215)

### Issues Resolved

BABEL-986

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 1]
```diff
@@ -493,6 +1327,19 @@ tsql_random(PG_FUNCTION_ARGS)
...
 }
 
+Datum
+timezone_mapping(PG_FUNCTION_ARGS)
+{
+	char *sqltmz = text_to_cstring(PG_GETARG_TEXT_P(0));
+	VarChar    *result = cstring_to_text("NULL");
+	for(int i=0;i<161;i++)
+	{
+		if(strcmp(win32_tzmap[i].stdname,sqltmz) == 0)
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/runtime/functions.c]: use pg_strcasecmp instead of strcmp

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 2]
```diff
@@ -493,6 +1327,19 @@ tsql_random(PG_FUNCTION_ARGS)
...
+		result = cstring_to_text(win32_tzmap[i].pgtzname);
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/runtime/functions.c]: add break once we find the match.

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 1 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Added

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 2 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Added

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 3]
```diff
@@ -982,6 +982,91 @@ $BODY$
...
+create or replace function sys.babelfish_timezone_mapping(IN tmz text) returns text
+AS 'babelfishpg_tsql', 'timezone_mapping'
+LANGUAGE C IMMUTABLE ;
+
+CREATE OR REPLACE FUNCTION sys.timezone(IN tzzone PG_CATALOG.TEXT ,  IN input_expr PG_CATALOG.TEXT)
+RETURNS sys.datetimeoffset
+AS
+$BODY$
+BEGIN
+    IF input_expr = 'NULL' THEN
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:994]: will last condition not cover this as well, why extra check for 'NULL'

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 4]
```diff
@@ -982,6 +982,91 @@ $BODY$
...
+    END IF;
+
+    IF pg_typeof(input_expr) IN ('sys.smalldatetime'::regtype, 'sys.datetime'::regtype, 'sys.datetime2'::regtype) THEN
+        input_expr_tx := input_expr::TEXT;
+        input_expr_tmz := input_expr_tx :: TIMESTAMPTZ;
+
+        result := (SELECT input_expr_tmz AT TIME ZONE tz_name)::TEXT;
+        tz_diff := (SELECT result::TIMESTAMPTZ - input_expr_tmz)::TEXT;
+        if LEFT(tz_diff,1) <> '-' THEN
+        tz_diff := concat('+',tz_diff);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: indentation

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 5]
```diff
@@ -1855,6 +1855,7 @@ class tsqlBuilder : public tsqlCommonMutator
 	//////////////////////////////////////////////////////////////////////////////
 	// function/procedure call analysis
 	//////////////////////////////////////////////////////////////////////////////
+
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: unnecessary changes

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 6]
```diff
@@ -982,6 +982,91 @@ $BODY$
...
+    tz_name := sys.babelfish_timezone_mapping(lower_tzn);
+    ELSE
+    tz_name := 'utc';
+    END IF;
+
+    IF tz_name = 'NULL' THEN
+        RAISE USING MESSAGE := format('Argument data type or the parameter %s provided to AT TIME ZONE clause is invalid.', tzzone);
+    END IF;
+
+    IF pg_typeof(input_expr) IN ('sys.smalldatetime'::regtype, 'sys.datetime'::regtype, 'sys.datetime2'::regtype) THEN
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:1038]: do we have test  cases for smalldatetime, datetime?
also negative test cases for what we do not support like time etc.

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 7]
```diff
@@ -179,6 +180,839 @@ const char *bbf_servicename = "MSSQLSERVER";
 char	   *bbf_language = "us_english";
 #define MD5_HASH_LEN 32
 
+
+static const struct
+{
+	const char *stdname;		/* Windows name of standard timezone */
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c]: indentation

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 8]
```diff
@@ -179,6 +180,839 @@ const char *bbf_servicename = "MSSQLSERVER";
...
+	const char *dstname;		/* Windows name of daylight timezone */
+	const char *pgtzname;		/* Name of pgsql timezone to map to */
+}			win32_tzmap[] =
+
+{
+	/*
+	 * This list was built from the contents of the registry at
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c]: let us verify if we are allowed to do this and should we comment the same as well

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 9]
```diff
@@ -493,6 +1327,22 @@ tsql_random(PG_FUNCTION_ARGS)
...
 }
 
+Datum
+timezone_mapping(PG_FUNCTION_ARGS)
+{
+	char *sqltmz = text_to_cstring(PG_GETARG_TEXT_P(0));
+	VarChar    *result = cstring_to_text("NULL");
+	for(int i=0;i<161;i++)
+	{
+		if(pg_strcasecmp(win32_tzmap[i].stdname,sqltmz) == 0)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c:507]: indentation and spacing

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 10]
```diff
@@ -493,6 +1327,22 @@ tsql_random(PG_FUNCTION_ARGS)
 	return result;
 }
 
+Datum
+timezone_mapping(PG_FUNCTION_ARGS)
+{
+	char *sqltmz = text_to_cstring(PG_GETARG_TEXT_P(0));
+	VarChar    *result = cstring_to_text("NULL");
+	for(int i=0;i<161;i++)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c]: use sizeof to get the size

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 11]
```diff
@@ -179,6 +180,839 @@ const char *bbf_servicename = "MSSQLSERVER";
 char	   *bbf_language = "us_english";
 #define MD5_HASH_LEN 32
 
+
+static const struct
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c]: put it in a separate header

File: contrib/babelfishpg_tsql/runtime/functions.c [hunk 12]
```diff
@@ -179,6 +180,839 @@ const char *bbf_servicename = "MSSQLSERVER";
...
+	 * This list was built from the contents of the registry at
+	 * HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time
+	 * Zones on Windows 7, Windows 10, and Windows Server 2019.  Some recent
+	 * additions have been made by comparing to the CLDR project's
+	 * windowsZones.xml file.
+	 *
+	 * The zones have been matched to IANA timezones based on CLDR's mapping
+	 * for "territory 001".
+	 */
+	{
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/runtime/functions.c]: as we are relying on postgres for core conversion, how are we making sure that every entry in the list works as expected for SQL server

Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Here in the mapping it is said that the list in taken from registry of windows which is the same directory as mentioned in tsql doc , so the mapping here are correct.
```
This list was built from the contents of the registry at HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Time Zones on Windows 7, Windows 10, and Windows Server 2019.  Some recent additions have been made by comparing to the CLDR project's windowsZones.xml file.
```

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 11 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Addressed

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 3 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:994]: Need to cover this case : -
```
 Select 'NULL' AT TIME ZONE NULL
 GO
```

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 4 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Addressed

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 6 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:1038]: Added

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 7 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Addressed

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 9 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c:507]: Addressed

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 10 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Done

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp (hunk 5 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Removed

File: test/JDBC/upgrade/latest/schedule [hunk 13]
```diff
@@ -10,6 +10,7 @@ BABEL-3613
 
 app_name
 atn2
+ATTIMEZONE-dep
```
Comment (forestkeeper) [inline test/JDBC/upgrade/latest/schedule:13]: should add ATTIMEZONE-dep to other schedule as well

Comment (hash-16) [reply test/JDBC/upgrade/latest/schedule:13]: This test file cannot be added to other schedule files of upgrade test because as the prepare runs on old version and verifies on the latest version so the prepare fails as it does not exist in older version.

**Review by forestkeeper — DISMISSED** [review]

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 12 above)
Comment (surendravishnoi) [reply contrib/babelfishpg_tsql/runtime/functions.c]: but that does not guarantee that for each entry BBF will give the same result as TSQL, have we tested these? Do we plan to add test for each entry?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 14]
```diff
@@ -982,6 +982,91 @@ $BODY$
...
+RETURNS sys.datetimeoffset
+AS
+$BODY$
+BEGIN
+    IF input_expr = 'NULL' THEN
+        RAISE USING MESSAGE := 'Argument data type varchar is invalid for argument 1 of AT TIME ZONE function.';
+    END IF;
+
+    IF input_expr IS NULL OR tzzone IS NULL THEN 
+    RETURN NULL;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: indentation

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 15]
```diff
@@ -982,6 +982,91 @@ $BODY$
...
+    lower_tzn PG_CATALOG.TEXT;
+    prev_res PG_CATALOG.TEXT;
+    result PG_CATALOG.TEXT;
+    is_dstt bool;
+    tz_diff PG_CATALOG.TEXT;
+    input_expr_tx PG_CATALOG.TEXT;
+    input_expr_tmz TIMESTAMPTZ;
+BEGIN
+    IF input_expr IS NULL OR tzzone IS NULL THEN 
+    RETURN NULL;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: indentation

File: contrib/babelfishpg_tsql/src/timezone.h [hunk 16]
```diff
@@ -0,0 +1,836 @@
+#ifndef TSQL_TIMEZONE_H
+#define TSQL_TIMEZONE_H
+
+static const struct
+{
+    const char *stdname;		/* Windows name of standard timezone */
+	const char *dstname;		/* Windows name of daylight timezone */
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/timezone.h:7]: indentation

File: contrib/babelfishpg_tsql/src/timezone.h [hunk 17]
```diff
@@ -0,0 +1,836 @@
...
+	const char *pgtzname;		/* Name of pgsql timezone to map to */
+}	win32_tzmap[] =
+
+{
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/timezone.h]: indentation

File: contrib/babelfishpg_tsql/runtime/functions.c (hunk 12 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/runtime/functions.c]: Will add test case for each timezone .

File: contrib/babelfishpg_tsql/src/timezone.h (hunk 17 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/src/timezone.h]: Addressed

File: contrib/babelfishpg_tsql/src/timezone.h (hunk 16 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/src/timezone.h:7]: Addressed

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 15 above)
Comment (hash-16) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql]: Addressed

**Review by forestkeeper — DISMISSED** [review]

**Review by surendravishnoi — DISMISSED** [review]

**Review by forestkeeper — APPROVED** [review]

**Review by surendravishnoi — APPROVED** [review]

**Review by basasairohan — APPROVED** [review]
